metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java

```diff
@@ -71,20 +76,36 @@ public void enrichment(RDF rdf) throws EnrichmentException {
   }
 
   @Override
-  public EnrichmentResultList enrichValues(List<InputValue> valuesForEnrichment)
+  public List<Pair<EnrichmentBase, EnrichmentFields>> enrichValues(
+      List<Pair<SearchValue, EnrichmentFields>> valuesForEnrichment)
       throws EnrichmentException {
     try {
-      return CollectionUtils.isEmpty(valuesForEnrichment) ? null
-          : retryableExternalRequestForNetworkExceptions(
-              () -> enrichmentClient.enrich(valuesForEnrichment));
+      if(CollectionUtils.isEmpty(valuesForEnrichment)){
+          return Collections.emptyList();
+      }
+      List<SearchValue> searchValues = valuesForEnrichment.stream().map(Pair::getFirst)
+          .collect(Collectors.toList());
+      EnrichmentResultList result = retryableExternalRequestForNetworkExceptions(
+          () -> enrichmentClient.enrich(searchValues));
+      List<List<EnrichmentBase>> enrichmentBaseLists = result.getEnrichmentBaseResultWrapperList()
+          .stream().map(EnrichmentResultBaseWrapper::getEnrichmentBaseList)
+          .collect(Collectors.toList());
+      List<Pair<EnrichmentBase, EnrichmentFields>> output = new ArrayList<>();
+      for (Pair<SearchValue, EnrichmentFields> pair : valuesForEnrichment) {
+        for (List<EnrichmentBase> enrichmentBases : enrichmentBaseLists) {
+          enrichmentBases.forEach(base -> output.add(new Pair<>(base, pair.getSecond())));
+        }
+      }
```

Comment:
Hmmm.... I think something goes wrong here. I'm not sure if I looked at this code before (and approved it), but I think this doesn't work. In this implementation you match every input with every output (so if there are n inputs, there will be n outputs and your loop goes through n*n iterations).

Instead you need to only go through n iterations.

Here is how I would do it:

```
      final List<Pair<EnrichmentBase, EnrichmentFields>> output = new ArrayList<>();
      for (int index = 0; index < searchValues.size(); index++){
        final EnrichmentFields fieldType = valuesForEnrichment.get(index).getSecond();
        result.getEnrichmentBaseResultWrapperList().get(index).getEnrichmentBaseList()
                .forEach(base -> output.add(new Pair<>(base, fieldType)));
      }
      return output;
```

Does this make sense?